[FEAT] 토큰 재발급 API 구현

dye0p opened the pull request:
## Issue Number
closed #253

## As-Is
<!-- 문제 상황 정의 -->
- 발급된 토큰의 만료시간이 지나면 로그인이 풀려, 사용자가 계속해서 로그인을 해줘야 하는 불편함이 있었음
- 문제를 해결하기 위해서 요청에서 토큰을 검증하여, 토큰이 만료되었다면 토큰을 재발급 해주는 API를 구현한다.

## To-Be
<!-- 변경 사항 -->
- 요청을 인터셉터로 가로채 토큰을 검증한다.
- 유효하지 않은 토큰의 경우 전역 예외에서 잡지 않고, 인터셉터에서 바로 401 UNAUTHORIZED 를 응답한다.
- 만료된 토큰의 경우 재발급 API를 통해서 재발급 요청을 할 수 있다.
- 토큰이 재발급 되면 동일하게 쿠키에 담아 응답한다.
- 기존 토큰 발급 로직은 토큰 재발급시 동일한 토큰이 재발급 되는 경우가 있었음. 동일한 토큰 재발급을 방지하기 위해서 토큰 발급시 UUID를 생성하여 토큰에 id를 부여하는 방법으로 동일한 토큰 발급을 방지하였다.


## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?

## (Optional) Additional Description
## merge 시기
해당 pr이 merge 되기 전, [[FEAT] 로그인 (인증/인가) 구현](https://github.com/woowacourse-teams/2025-Fit-toring/pull/252) PR이 먼저 merge 되어야 합니다.

## 인터셉터 동작 엔드포인트

우선, 인터셉터가 동작하는 경로를 아직 지정하지 않고, 모든 경로에 대해서 인터셉터가 동작하지 않도록 하였습니다.
이 부분은 팀내 회의를 통해서 지정해야 할 것 같습니다.

인터셉터가 동작하는 엔드포인트가 정해지면, 쿠키 추가를 위해서 기존의 통합 테스트가 수정되어야 할 것 같습니다.

dye0p committed:
feat: AuthenticationInterceptor 뼈대 코드 준비

- 토큰이 유효한지 검증 해야하기 떄문에 Http 요청에서 인증 정보를 먼저 검증해야한다.
- 토큰이 쿠키에 담겨있기 때문에 Http 요청을 가로채야 한다.

dye0p committed:
feat: 인터셉터에서 가로챈 요청에 Cookie가 존재하는지 검증 하고, 예외가 발생하면 예외를 응답하는 기능 구현

dye0p committed:
feat: Cookie에서 토큰을 추출하고, 예외가 발생하면 예외를 응답하는 기능 구현

dye0p committed:
feat: refreshToken으로 토큰을 재발급 하는 기능 구현

- refreshToken을 find하는 과정에서 발생하는 예외를 전역 예외 핸들러에서 잡도록 추가
- 새로운 accessToken과 refreshToken을 발급한다.

dye0p committed:
feat: Token을 발급 할 때 UUID를 가지는 (JWT ID) 추가

- 토큰 재발급시 동일한 토큰이 발급되는 문제가 있었음
- 동일한 토큰 재발급 방지를 위해서 토큰발급 마다 랜덤으로 생성되는 UUID를 추가하여 매번 다른 토큰이 발급되도록 보장할 수 있다.

dye0p committed:
test: authService 토큰 재발급 테스트 추가

dye0p committed:
feat: 토큰 재발급 API 구현

dye0p committed:
feat: 인터셉터 WebConfig에 등록

dye0p committed:
feat: 인터셉터 검증에서 엔드포인트 설정

- 검증에서 제외될 엔드포인트는 excludePathPatterns 메서드에 명시하였음.

kimdevv commented on backend/fittoring/src/main/java/fittoring/config/auth/AuthenticationInterceptor.java:
```diff
@@ -0,0 +1,48 @@
+        }
+
+        try {
+            String accessToken = jwtExtractor.extractTokenFromCookie("accessToken", cookies);
```
p3: 액세스 토큰을 저장할 키 이름(지금 `accessToken`)도 논의해 봐야겠네여

kimdevv commented on backend/fittoring/src/main/java/fittoring/config/auth/AuthenticationInterceptor.java:
```diff
@@ -0,0 +1,48 @@
+import org.springframework.web.servlet.HandlerInterceptor;
+
+@RequiredArgsConstructor
+@Component
+public class AuthenticationInterceptor implements HandlerInterceptor {
```
p3: 이 친구는 뭔가 `@Component`가 없어도 되고 있어도 될 거 같은데 쓰다 보니 사실 아무 차이도 없는 거 같긴 하네요

dye0p replied:
```java
@RequiredArgsConstructor
@Configuration
public class WebConfig implements WebMvcConfigurer {

    private final AuthenticationInterceptor authenticationInterceptor;

    ...

    @Override
    public void addInterceptors(InterceptorRegistry registry) {
        registry.addInterceptor(authenticationInterceptor)
                .excludePathPatterns("/**");
    }
}
```

webConfig 파일에 등록하려면 `AuthenticationInterceptor`객체가 컴포넌트 스캔의 대상이 되어야 합니다! 
따라서 필요할 것 같아요~

kimdevv commented on backend/fittoring/src/main/java/fittoring/config/auth/AuthenticationInterceptor.java:
```diff
@@ -0,0 +1,48 @@
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
+            throws Exception {
+        Cookie[] cookies = request.getCookies();
```
p100: 내가 만든 쿠키~

ppparkta replied:
??

dye0p replied:
코드리뷰는 신성한 공간 입니다. 자중해주세요.

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/RefreshToken.java:
```diff
@@ -30,10 +30,16 @@ public class RefreshToken {
     @Column(nullable = false)
     private String token;
```
p3: `token`도 직관적이라 좋은 이름이긴 한데 뭔가 `RefreshToken.token`이라고 하면 토큰의 토큰? 같은 느낌이라서, `value`라는 네이밍은 어떤가요???

ppparkta replied:
조은데요 근데 엔티티 내에 다른 필드가 생각보다 많아서 value라고만 하면 그것대로 혼란스러울 수 있을 것 같아요. RefreshToken.tokenValue 정도면 단일 필드만으로도 의미 유추가 가능해서 좋을거 같기도 하고요

dye0p replied:
tokenValue로 변경하겠습니다!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/AuthService.java:
```diff
@@ -47,6 +48,22 @@ public AuthTokenResponse login(String loginId, String password) {
+    public AuthTokenResponse reissue(String refreshToken) {
+        jwtProvider.validateToken(refreshToken);
+        RefreshToken findRefreshToken = getRefreshToken(refreshToken);
+        String newAccessToken = jwtProvider.createToken(findRefreshToken.getMemberId());
```
p3: 액세스 토큰을 만들 멤버의 식별자가 리프레시 토큰이랑 같이 저장되어 있어야 해서 DB에 저장되어야 하는 거군요

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/JwtExtractor.java:
```diff
@@ -0,0 +1,19 @@
+@Component
+public class JwtExtractor {
+
+    public String extractTokenFromCookie(String name, Cookie[] cookies) {
+        return Arrays.stream(cookies)
+                .filter(cookie -> cookie.getName().equals(name))
+                .map(Cookie::getValue)
+                .findFirst()
+                .orElseThrow(() -> new InvalidTokenException(BusinessErrorMessage.NOT_FOUND_TOKEN.getMessage()));
+    }
```
p3: 쿠키에서 토큰을 추출할 수도 있지만 뭔가 `name` 매개변수에 따라서 다양한 값을 추출할 수도 있을 것 같아요. `CookieWriter`이라는 클래스가 있었던 것 같은데 그 클래스 이름을 `CookieManager`으로 바꾸고 이 메서드를 거기로 옮기는 건 어떤가요??? 예외 메세지도 좀 더 범용적으로 바꾸고요! ('존재하지 않는 키 값입니다.' 같은?)

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/AuthController.java:
```diff
@@ -36,15 +40,31 @@ public ResponseEntity<Void> login(@RequestBody @Valid SignInRequest request, Htt
 
+    @PostMapping("/reissue")
+    public ResponseEntity<Void> reissue(
+            @CookieValue(REFRESH_TOKEN_COOKIE_NAME) String refreshToken,
```
p3: 이렇게 하면 쿠키 값도 바로 가져올 수 있군요 ㄷㄷ

kimdevv reviewed (approved):
수고하셨습니다~~

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/RefreshToken.java:
```diff
@@ -30,10 +30,16 @@ public class RefreshToken {
+    public void update(String token, LocalDateTime createAt) {
+        this.token = token;
+        this.createAt = createAt;
+    }
```
p3: 캡슐화 좋네요. 근데 궁금한게, refresh token을 매번 같은걸 갱신만 해서 쓰나요? 완전히 별개의 토큰값을 새로 생성해주기 때문에 아예 새로운 레코드로 만드는건 별로일까요?

ppparkta replied:
코드설명회에서 해결되었습니다.

ppparkta reviewed (approved):
코드 설명회 유익했습니다. 쿠키 만료 시간을 추가하면 더 완벽해지겠네요. 고생많으셨습니다.

dye0p committed:
Merge branch 'feature/173-feat-signup' into feature/253-reissue-token

dye0p committed:
fix: 잘못된 병합 수정

dye0p committed:
refactor: RefreshToken 엔티티 token 필드명 명확하게 변경(token -> tokenValue)

dye0p committed:
refactor: RefreshTokenRepository 상속 객체 변경
(JpaRepository -> ListCrudRepository)

- 아직까지 JpaRepository의 기능을 사용하지 않음.
- ListCrudRepository로 충분히 가능하기에 변경
